File: code/平衡二叉树/avl.js
/**
 * @description  创建树，新增节点，删除节点 代码演示
 * @author 氧化氢
 */


// 树节点的数据结构：
function Node(val, left, right, h = 1) {
  this.val = val ? val : -1000
  this.left = left ? left : null
  this.right = right ? right : null
  this.h = (typeof h === 'number') ? h : 1
}


function getNewNode(val) {
  return new Node(val, null, null, 1)
}

// 删除整个树中 root 之后的节点
function clear(root) {
  if(!root) return
  clear(root.left)
  clear(root.right)
  root.left = null
  root.right = null
  root.val = -10000

  return
}


// 插入：
function insert(root, key) {
  if(!root) return getNewNode(key)
  if(root.val === key) return root
  if(key < root.val) {
    root.left = insert(root.left, key)
  }else {
    root.right = insert(root.right, key)
  } 
  update_height(root)
  return maintain(root)
}


// 更新树高：
function update_height(root) {
  // 左右子树都没有，那么树高就是默认的 1
  if(root.left === null && root.right === null) return
  // 左子树为空的时候，树高就等于右子树树高 + 1
  if(root.left === null) {
    root.h = root.right.h + 1
  }
  // 右子树为空时，树高就等于左子树树高 + 1
  else if(root.right === null) {
    root.h = root.left.h + 1
  }
  // 左右子树都存在时，树高就等于最大树高 + 1
  else root.h = Math.max(root.left.h, root.right.h) + 1
}



// 删除节点：
function deleteNode(root, key) {
  if(!root) return root
  if(key < root.val) {
    root.left = deleteNode(root.left, key)
  }else if(key > root.val) {
    root.right = deleteNode(root.right, key)
  }else {
    // 度为 0
    if(!root.left && !root.right) {
      // delete root
      return null
    }
    // 度为 1
    else if(!root.left || !root.right) {
      var temp = root.left ? root.left : root.right
      // delete root
      return temp
    }
    // 度为 2
    else {
      var temp = predeccessor(root)
      root.val = temp.val
      root.left = deleteNode(root.left, temp.val)
    }
  }

  update_height(root)

  return maintain(root)
}

// 找到前驱节点：
function predeccessor(root) {
  var temp = root.left
  while(temp.right) temp = temp.right

  return temp
}

// 左旋：
function left_rotate(root) {
  var newRoot = root.right
  root.right = newRoot.left
  newRoot.left = root
  update_height(root)
  update_height(newRoot)
  return newRoot
}

function right_rotate(root) {
  var newRoot = root.left
  root.left = newRoot.right
  newRoot.right = root
  update_height(root)
  update_height(newRoot)
  return newRoot
}

function LR_rotate(root) {
  var temp = root.left.right
  temp.left = root.left
  temp.right = root
}
function RL_rotate(root) {
  var temp = root.right.left
  temp.right = root.right
  temp.left = root
}

// 调整节点，调整平衡：
function maintain(root) {

  if(!root.left && !root.right) return root
  if(root.left && root.left.h === 1 && !root.right) return root
  if(root.right && root.right.h === 1 && !root.left) return root
  if(root.left && root.left.left && root.left.h === 2 && !root.right) return right_rotate(root)
  if(root.left && root.left.right && root.left.h === 2 && !root.right) return LR_rotate(root)
  if(root.right && root.right.right && root.right.h === 2 && !root.left) return left_rotate(root)
  if(root.right && root.right.left && root.right.h === 2 && !root.left) return RL_rotate(root)
  if(root.left && root.left.h === 2 && root.right.h === 1) return root
  if(root.right && root.right.h === 2 && root.left.h === 1) return root


  // root 左右至少都有两个节点的情况下可以正常执行:
  if(Math.abs(root.left.h - root.right.h) < 2) return root
  if(root.left.h > root.right.h) {
    if(root.left.right.h > root.left.left.h) {
      // LR 的情况：先左旋，再右旋
      root.left = left_rotate(root.left)
    }
    // 不走上面 if 语句就是 LL 情况，直接右旋即可
    root = right_rotate(root)
  } else {
    if(root.right.left.h > root.right.right.h) {
      // RL 的情况：先左旋，再右旋
      root.right = right_rotate(root.right)
    }
    // 不走上面 if 语句就是 RR 情况，直接左旋即可
    root = left_rotate(root) 
  }

  return root
}


// 测试代码：
function test() {
  // RR型测试：
  // var tree = new Node(1, null, null, 1)
  // tree = insert(tree, 2)
  // tree = insert(tree, 3)
  // tree = insert(tree, 4)
  // tree = insert(tree, 5)
  // tree = insert(tree, 6)
  // tree = insert(tree, 7)
  // tree = insert(tree, 8)


  // LL 型测试
  // var tree = new Node(8, null, null, 1)
  // tree = insert(tree, 7)
  // tree = insert(tree, 6)
  // tree = insert(tree, 5)
  // tree = insert(tree, 4)
  // tree = insert(tree, 3)
  // tree = insert(tree, 2)
  // tree = insert(tree, 1)


  // LR 型测试
  // var tree = new Node(8, null, null, 1)
  // tree = insert(tree, 9)
  // tree = insert(tree, 4)
  // tree = insert(tree, 3)
  // tree = insert(tree, 5)
  // tree = insert(tree, 6)


  // RL 型测试
  var tree = new Node(4, null, null, 1)
  tree = insert(tree, 3)
  tree = insert(tree, 8)
  tree = insert(tree, 9)
  tree = insert(tree, 5)
  tree = insert(tree, 6)
  console.log(tree);

  // console.log(tree);

  // insert(tree, 4)
  // insert(tree, 5)


  // console.log('插入后的树：', tree);


  // deleteNode(tree, 3)
  // deleteNode(tree, 4)

  // console.log('删除后的树：', tree);

  // clear(tree)

  // console.log(tree);


  var arr = []
  midOrder(tree, arr)

  console.log(arr);

  return arr
}

// 中序遍历：
function midOrder(root, arr) {
  if(!root) return
  midOrder(root.left, arr)
  if(root.val !== -1000) arr.push(root.val)
  midOrder(root.right, arr) 
}


test()

File: code/平衡二叉树/avl.py
"""
平衡二叉树
"""


class TreeNode:
    """
    节点
    """

    def __init__(self, val, h, left, right,):
        """
        树中节点的值，节点的树高，节点的左孩子和有孩子
        """
        self.val = val
        self.h = h
        self.left = left
        self.right = right


class Node:
    """
    虚拟节点，替代空节点，将值为 -999 的节点设置为虚拟节点
    """

    def __init__(self, val=-999, h=0):
        self.val = val
        self.h = h


class Main:
    def getPreOrder(self, root):
        pre_order = []
        od = Order()
        od.pre_order(root, pre_order)

        return pre_order

    def getMidOrder(self, root):
        mid_order = []
        od = Order()
        od.mid_order(root, mid_order)

        return mid_order


class Order:
    def pre_order(self, root, arr):
        if root.val == -999:
            return
        self.my_print(root)
        arr.append(root.val)
        self.pre_order(root.left, arr)
        self.pre_order(root.right, arr)

    def mid_order(self, root, arr):
        if root.val == -999:
            return
        self.mid_order(root.left, arr)
        arr.append(root.val)
        self.mid_order(root.right, arr)

    def my_print(self, root):
        print(f"({root.val}[{root.h}] | {root.left.val}, {root.right.val})\n")
        return


class TreeOperate:

    def getNewNode(self, key):
        """
        得到值是 key 的一个新的树节点
        """
        return TreeNode(val=key, h=1, left=Node(), right=Node())

    def clear(self, root):
        """
        删除当前树
        """
        if root.val == -999:
            return
        self.clear(root.left)
        self.clear(root.right)
        # 删除当前节点
        print(f"节点 {root.val} 被删除！")
        # TODO: 删除当前节点
        # root = Node()
        root = None
        return root

    def getPreNode(self, root):
        """
        获得当前节点的前驱节点，这里注意前驱节点不是父节点，而是中序遍历时当前节点的前面节点，注意是中序遍历
        """
        temp = root.left
        while temp.right.val != -999:
            temp = temp.right

        return temp

    def update_height(self, root):
        if root.left is None and root.right is None:
            return
        if root.left is None:
            root.h = root.right.h + 1
        elif root.right is None:
            root.h = root.left.h + 1
        else:
            root.h = max(root.left.h, root.right.h) + 1

    def maintain(self, root):
        """
        对从 root 节点出发的不平衡的树进行调整，返回调整平衡后的二叉树
        """
        if abs(root.left.h - root.right.h) < 2:
            return root
        """
        四种失衡状态：LL 1、LR 2、RL 3、RR 4
        LR：先小左旋，后大右旋
        RL：先小右旋，后大左旋
        """
        type = -1
        if root.left.h > root.right.h:
            # 第一个字母是 L
            if root.left.right.h > root.left.left.h:
                # 第二个字母是 R，失衡类型是 LR，先小左旋，再大右旋
                # 小左旋
                if type == -1:
                    type = 2
                root.left = self.left_rotate(root.left)
            # 对根节点大右旋
            if type == -1:
                type = 1
            root = self.right_rotate(root)
        else:
            if root.right.left.h > root.right.right.h:
                # RL 类型，先小右旋，再大左旋
                if type == -1:
                    type = 3
                root.right = self.right_rotate(root.right)
            # 对根节点大左旋
            if type == -1:
                type = 4
            root = self.left_rotate(root)

        # 输入当前树失衡类型
        str = ''
        if type == 1:
            str = 'LL'
        elif type == 2:
            str = 'LR'
        elif type == 3:
            str = 'RL'
        else:
            str = 'RR'
        print(f'\033[32m当前树不平衡的类型为：{str}！\033[0m')

        return root

    def left_rotate(self, root):
        new_root = root.right
        root.right = new_root.left
        new_root.left = root
        self.update_height(root)
        self.update_height(new_root)
        return new_root

    def right_rotate(self, root):
        new_root = root.left
        root.left = new_root.right
        new_root.right = root
        self.update_height(root)
        self.update_height(new_root)
        return new_root

    def insert(self, root, key):
        """
        插入值为 key 的节点，返回插入后的树的根节点的地址
        """
        if root.val == -999:
            return self.getNewNode(key)
        # 防止重复插入
        if root.val == key:
            return root
        if root.val > key:
            root.left = self.insert(root.left, key)
        if root.val < key:
            root.right = self.insert(root.right, key)

        # 更新当前节点的高度
        self.update_height(root)

        return self.maintain(root)

    def erase(self, root, key):
        """
        二叉排序树的删除操作，删除单个节点
        """
        if root.val == -999:
            return root
        if root.val > key:
            root.left = self.erase(root.left, key)
        elif root.val < key:
            root.right = self.erase(root.right, key)
        else:
            # 要删除的节点就是当前节点：
            # if root.left is None and root.right is None:
            #     # 删除度为 0 的节点
            #     root = None
            #     return None
            if root.left.val == -999 or root.right.val == -999:
                # 删除度 为 0 和 1 的节点
                temp = root.right if root.left.val == -999 else root.left
                # TODO: 删除当前节点
                # root = Node()
                root = None
                return temp
            else:
                # 删除度为 2 的节点
                temp = self.getPreNode(root)
                root.val = temp.val
                root.left = self.erase(root.left, temp.val)
        self.update_height(root)

        return self.maintain(root)


# test 程序:
tree = Node()  # 初始化空树
treeOp = TreeOperate()
m = Main()
while 1:
    op, val = (input('请输入操作符和节点值，注意用空格分开：').split())
    op = int(op)
    val = int(val)
    print(f'操作符：{op}，节点值：{val}')
    if op == -1:
        print('退出操作！')
        break
    if op == 0:
        # 执行插入操作
        tree = treeOp.insert(tree, val)
    if op == 1:
        # 执行删除操作
        tree = treeOp.erase(tree, val)

    print("\033[34m====== AVL Tree Print =======\n\033[0m")
    print("格式：树值[树高] | 左子树的值，右子树的值")
    m.getPreOrder(tree)
    # print(f'树的中序遍历：{arr}\n')
    print("\033[34m====== Tree Print Down =======\n\033[0m")
